File: python/gestor_sga.py
from __future__ import annotations

import os
from typing import Dict, List

from data_structures import Cola, Pila, RegistroNota
from models import Alumno, Profesor, ProgramaAcademico, crear_programa


class GestorSGA:
    ARCHIVO_ALUMNOS = "alumnos.txt"
    ARCHIVO_PROFESORES = "profesores.txt"
    ARCHIVO_CERTIFICADOS = "certificados_pendientes.txt"

    def __init__(self, directorio_datos: str = ".") -> None:
        self._directorio_datos: str = directorio_datos
        self._alumnos: Dict[str, Alumno] = {}
        self._profesores: Dict[str, Profesor] = {}
        self._pila_notas: Pila[RegistroNota] = Pila()
        self._cargar_datos()

    def _ruta(self, nombre_archivo: str) -> str:
        return os.path.join(self._directorio_datos, nombre_archivo)


    # Registro de personas
    def registrar_alumno(self, cedula: str, nombre: str, correo: str, tipo_programa: str) -> Alumno:
        if not cedula or not nombre:
            raise ValueError("La cédula y el nombre del alumno son obligatorios.")
        if cedula in self._alumnos:
            raise ValueError(f"Ya existe un alumno registrado con la cédula {cedula}.")
        programa: ProgramaAcademico = crear_programa(tipo_programa)
        alumno = Alumno(cedula, nombre, correo, programa)
        self._alumnos[cedula] = alumno
        return alumno

    def registrar_profesor(
        self, cedula: str, nombre: str, correo: str, especialidad: str, materia: str
    ) -> Profesor:
        if not cedula or not nombre:
            raise ValueError("La cédula y el nombre del profesor son obligatorios.")
        if cedula in self._profesores:
            raise ValueError(f"Ya existe un profesor registrado con la cédula {cedula}.")
        profesor = Profesor(cedula, nombre, correo, especialidad, materia)
        self._profesores[cedula] = profesor
        return profesor


    # Notas + Pila LIFO
    def registrar_nota(self, cedula: str, nota: float) -> None:
        alumno = self._alumnos.get(cedula)
        if alumno is None:
            raise ValueError(f"No existe un alumno registrado con la cédula {cedula}.")
        alumno.agregar_nota(nota)
        self._pila_notas.apilar((cedula, nota))

    def deshacer_ultima_nota(self) -> RegistroNota:
        """Desapila el último registro de nota y lo elimina del alumno correspondiente."""
        cedula, nota = self._pila_notas.desapilar()
        alumno = self._alumnos.get(cedula)
        if alumno is None:
            raise ValueError("El alumno asociado a ese registro ya no existe en el sistema.")
        alumno.eliminar_ultima_nota()
        return cedula, nota


    # Cola FIFO (Certificados)
    def generar_cola_certificados(self) -> List[Alumno]:
        cola: Cola[Alumno] = Cola()
        for alumno in self._alumnos.values():
            if alumno.esta_aprobado():
                cola.encolar(alumno)

        procesados: List[Alumno] = []
        while not cola.esta_vacia():
            procesados.append(cola.desencolar())

        self._exportar_certificados(procesados)
        return procesados

    def _exportar_certificados(self, alumnos_aprobados: List[Alumno]) -> None:
        with open(self._ruta(self.ARCHIVO_CERTIFICADOS), "w", encoding="utf-8") as archivo:
            archivo.write("=== REPORTE DE CERTIFICADOS PENDIENTES (procesado en orden FIFO) ===\n\n")
            if not alumnos_aprobados:
                archivo.write("No hay alumnos aprobados pendientes de certificación.\n")
                return
            for posicion, alumno in enumerate(alumnos_aprobados, start=1):
                archivo.write(
                    f"{posicion}. Cédula: {alumno.get_cedula()} | Nombre: {alumno.get_nombre()} "
                    f"| Programa: {alumno.get_programa()} | Promedio: {alumno.get_promedio():.2f}\n"
                )


    # Reportes
    def obtener_reporte_general(self) -> str:
        lineas: List[str] = ["=== ALUMNOS ==="]
        if not self._alumnos:
            lineas.append("No hay alumnos registrados.")
        for alumno in self._alumnos.values():
            lineas.append(alumno.to_string())

        lineas.append("\n=== PROFESORES ===")
        if not self._profesores:
            lineas.append("No hay profesores registrados.")
        for profesor in self._profesores.values():
            lineas.append(profesor.to_string())

        return "\n".join(lineas)

    def listar_alumnos(self) -> List[Alumno]:
        return list(self._alumnos.values())

    def listar_profesores(self) -> List[Profesor]:
        return list(self._profesores.values())


    # Persistencia en archivos .txt
    def guardar_datos(self) -> None:
        self._guardar_alumnos()
        self._guardar_profesores()

    def _guardar_alumnos(self) -> None:
        with open(self._ruta(self.ARCHIVO_ALUMNOS), "w", encoding="utf-8") as archivo:
            for alumno in self._alumnos.values():
                notas = alumno.get_notas()
                notas_completas = notas + [0] * (Alumno.MAX_NOTAS - len(notas))
                notas_str = ",".join(str(nota) for nota in notas_completas[: Alumno.MAX_NOTAS])
                tipo_programa = alumno.get_programa().get_nombre_programa()
                linea = (
                    f"{alumno.get_cedula()},{alumno.get_nombre()},{alumno.get_correo()},"
                    f"{tipo_programa},{notas_str}\n"
                )
                archivo.write(linea)

    def _guardar_profesores(self) -> None:
        with open(self._ruta(self.ARCHIVO_PROFESORES), "w", encoding="utf-8") as archivo:
            for profesor in self._profesores.values():
                linea = (
                    f"{profesor.get_cedula()},{profesor.get_nombre()},{profesor.get_correo()},"
                    f"{profesor.get_especialidad()},{profesor.get_materia()}\n"
                )
                archivo.write(linea)

    def _cargar_datos(self) -> None:
        self._cargar_alumnos()
        self._cargar_profesores()

    def _cargar_alumnos(self) -> None:
        ruta = self._ruta(self.ARCHIVO_ALUMNOS)
        if not os.path.exists(ruta):
            return
        with open(ruta, "r", encoding="utf-8") as archivo:
            for linea in archivo:
                linea = linea.strip()
                if not linea:
                    continue
                try:
                    partes = linea.split(",")
                    cedula, nombre, correo, tipo_programa = partes[0], partes[1], partes[2], partes[3]
                    notas_crudas = partes[4 : 4 + Alumno.MAX_NOTAS]
                    programa = crear_programa(tipo_programa)
                    alumno = Alumno(cedula, nombre, correo, programa)
                    for nota_texto in notas_crudas:
                        valor = float(nota_texto)
                        if valor > 0:
                            alumno.agregar_nota(valor)
                    self._alumnos[cedula] = alumno
                except (ValueError, IndexError):
                    continue

    def _cargar_profesores(self) -> None:
        ruta = self._ruta(self.ARCHIVO_PROFESORES)
        if not os.path.exists(ruta):
            return
        with open(ruta, "r", encoding="utf-8") as archivo:
            for linea in archivo:
                linea = linea.strip()
                if not linea:
                    continue
                try:
                    cedula, nombre, correo, especialidad, materia = linea.split(",")
                    self._profesores[cedula] = Profesor(cedula, nombre, correo, especialidad, materia)
                except (ValueError, IndexError):
                    continue

File: java/src/com/diplomadosonline/sga/services/GestorSGA.java
package com.diplomadosonline.sga.services;

import com.diplomadosonline.sga.models.Alumno;
import com.diplomadosonline.sga.models.Bootcamp;
import com.diplomadosonline.sga.models.Curso;
import com.diplomadosonline.sga.models.Diplomado;
import com.diplomadosonline.sga.models.Profesor;
import com.diplomadosonline.sga.models.ProgramaAcademico;
import com.diplomadosonline.sga.models.RegistroNota;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;
import java.util.Queue;
import java.util.Stack;

public class GestorSGA {

    private static final String DIRECTORIO_DATOS = "java";

    private static final String ARCHIVO_ALUMNOS = "alumnos.txt";
    private static final String ARCHIVO_PROFESORES = "profesores.txt";
    private static final String ARCHIVO_CERTIFICADOS = "certificados_pendientes.txt";

    private final List<Alumno> alumnos;
    private final List<Profesor> profesores;

    // Pila (LIFO) 
    private final Stack<RegistroNota> pilaNotas;
    
    public GestorSGA() {
        this.alumnos = new ArrayList<>();
        this.profesores = new ArrayList<>();
        this.pilaNotas = new Stack<>();
    }

    // Registro de personas
    public void registrarAlumno(Alumno alumno) {
        alumnos.add(alumno);
    }

    public void registrarProfesor(Profesor profesor) {
        profesores.add(profesor);
    }

    public Optional<Alumno> buscarAlumnoPorCedula(String cedula) {
        return alumnos.stream()
                .filter(a -> a.getCedula().equalsIgnoreCase(cedula))
                .findFirst();
    }

    public List<Alumno> getAlumnos() {
        return new ArrayList<>(alumnos);
    }

    public List<Profesor> getProfesores() {
        return new ArrayList<>(profesores);
    }

    // Pila (Stack) LIFO — Deshacer nota
    public void registrarNotaAlumno(String cedula, double nota) {
        Alumno alumno = buscarAlumnoPorCedula(cedula)
                .orElseThrow(() -> new IllegalArgumentException("No existe un alumno con cédula: " + cedula));
        alumno.agregarNota(nota);
        pilaNotas.push(new RegistroNota(cedula, nota));
    }

    public Optional<RegistroNota> deshacerUltimoRegistroNota() {
        if (pilaNotas.isEmpty()) {
            return Optional.empty();
        }
        RegistroNota ultimo = pilaNotas.pop();
        buscarAlumnoPorCedula(ultimo.getCedulaAlumno()).ifPresent(Alumno::eliminarUltimaNota);
        return Optional.of(ultimo);
    }

    // Cola (Queue) FIFO — Certificados
    public int generarColaCertificados() throws IOException {
        Queue<Alumno> colaCertificados = new LinkedList<>();

        for (Alumno alumno : alumnos) {
            if (alumno.esAprobado()) {
                colaCertificados.offer(alumno);
            }
        }

        List<String> lineasReporte = new ArrayList<>();
        lineasReporte.add("=== REPORTE DE CERTIFICADOS PENDIENTES (Orden FIFO) ===");
        int contador = 0;

        Alumno actual;
        while ((actual = colaCertificados.poll()) != null) {
            contador++;
            String nombrePrograma = (actual.getPrograma() != null)
                    ? actual.getPrograma().getNombrePrograma()
                    : "N/A";
            lineasReporte.add(String.format("%d) Cedula: %s | Nombre: %s | Programa: %s | Notas: %s",
                    contador, actual.getCedula(), actual.getNombre(), nombrePrograma, actual.getNotas()));
        }

        if (contador == 0) {
            lineasReporte.add("No hay alumnos aprobados pendientes de certificación.");
        }

        escribirLineas(ARCHIVO_CERTIFICADOS, lineasReporte);
        return contador;
    }

    // Reporte general
    public String generarReporteGeneral() {
        StringBuilder sb = new StringBuilder();

        sb.append("=== ALUMNOS ===\n");
        if (alumnos.isEmpty()) {
            sb.append("(sin alumnos registrados)\n");
        } else {
            for (Alumno a : alumnos) {
                String programa = (a.getPrograma() != null)
                        ? capitalizar(a.getPrograma().getTipo())
                        : "N/A";
                String estado = a.esAprobado() ? "APROBADO" : "NO APROBADO";
                sb.append(String.format(
                        "Cédula: %s | Nombre: %s | Correo: %s | Programa: %s | Notas: %s | Promedio: %.2f | Estado: %s%n",
                        a.getCedula(), a.getNombre(), a.getCorreo(), programa,
                        a.getNotas(), a.getPromedio(), estado));
            }
        }

        sb.append('\n');
        sb.append("=== PROFESORES ===\n");
        if (profesores.isEmpty()) {
            sb.append("(sin profesores registrados)\n");
        } else {
            for (Profesor p : profesores) {
                sb.append(String.format(
                        "Cédula: %s | Nombre: %s | Correo: %s | Especialidad: %s | Materia: %s%n",
                        p.getCedula(), p.getNombre(), p.getCorreo(), p.getEspecialidad(), p.getMateria()));
            }
        }

        return sb.toString();
    }

    private static String capitalizar(String texto) {
        if (texto == null || texto.isEmpty()) {
            return texto;
        }
        return texto.substring(0, 1).toUpperCase() + texto.substring(1).toLowerCase();
    }


    // Persistencia
    public void cargarDatos() throws IOException {
        cargarAlumnos();
        cargarProfesores();
    }

    private Path resolverRutaDatos(String nombreArchivo) throws IOException {
        Path carpeta = Paths.get(DIRECTORIO_DATOS);
        if (!Files.exists(carpeta)) {
            Files.createDirectories(carpeta);
        }
        return carpeta.resolve(nombreArchivo);
    }

    private void cargarAlumnos() throws IOException {
        Path ruta = resolverRutaDatos(ARCHIVO_ALUMNOS);
        if (!Files.exists(ruta)) {
            return;
        }
        try (BufferedReader reader = Files.newBufferedReader(ruta)) {
            String linea;
            while ((linea = reader.readLine()) != null) {
                if (linea.isBlank()) {
                    continue;
                }
                String[] campos = linea.split(",", -1);
                if (campos.length < 7) {
                    continue; 
                }
                String cedula = campos[0].trim();
                String nombre = campos[1].trim();
                String correo = campos[2].trim();
                String tipoPrograma = campos[3].trim();

                ProgramaAcademico programa = crearProgramaPorTipo(tipoPrograma);

                Alumno alumno = new Alumno(cedula, nombre, correo, programa);
                List<Double> notas = new ArrayList<>();
                try {
                    double n1 = Double.parseDouble(campos[4].trim());
                    double n2 = Double.parseDouble(campos[5].trim());
                    double n3 = Double.parseDouble(campos[6].trim());
                    if (n1 > 0) notas.add(n1);
                    if (n2 > 0) notas.add(n2);
                    if (n3 > 0) notas.add(n3);
                } catch (NumberFormatException e) {
                }
                alumno.setNotas(notas);
                alumnos.add(alumno);
            }
        }
    }

    private void cargarProfesores() throws IOException {
        Path ruta = resolverRutaDatos(ARCHIVO_PROFESORES);
        if (!Files.exists(ruta)) {
            return;
        }
        try (BufferedReader reader = Files.newBufferedReader(ruta)) {
            String linea;
            while ((linea = reader.readLine()) != null) {
                if (linea.isBlank()) {
                    continue;
                }
                String[] campos = linea.split(",", -1);
                if (campos.length < 5) {
                    continue;
                }
                Profesor profesor = new Profesor(
                        campos[0].trim(),
                        campos[1].trim(),
                        campos[2].trim(),
                        campos[3].trim(),
                        campos[4].trim());
                profesores.add(profesor);
            }
        }
    }

    private ProgramaAcademico crearProgramaPorTipo(String tipo) {
        switch (tipo.toUpperCase()) {
            case "DIPLOMADO":
                return new Diplomado("Diplomado");
            case "BOOTCAMP":
                return new Bootcamp("Bootcamp");
            case "CURSO":
                return new Curso("Curso");
            default:
                return new Curso("Curso");
        }
    }

    public void guardarDatos() throws IOException {
        guardarAlumnos();
        guardarProfesores();
    }

    private void guardarAlumnos() throws IOException {
        List<String> lineas = new ArrayList<>();
        for (Alumno a : alumnos) {
            lineas.add(a.toCSV());
        }
        escribirLineas(ARCHIVO_ALUMNOS, lineas);
    }

    private void guardarProfesores() throws IOException {
        List<String> lineas = new ArrayList<>();
        for (Profesor p : profesores) {
            lineas.add(p.toCSV());
        }
        escribirLineas(ARCHIVO_PROFESORES, lineas);
    }

    private void escribirLineas(String nombreArchivo, List<String> lineas) throws IOException {
        Path ruta = resolverRutaDatos(nombreArchivo);
        try (BufferedWriter writer = Files.newBufferedWriter(ruta)) {
            for (String linea : lineas) {
                writer.write(linea);
                writer.newLine();
            }
        }
    }
}
